Resilient Backpropagation Optimizer

ciprianflow opened the pull request:
Rprop was originally proposed by Riedmiller and Braun in the paper [A direct adaptive method for faster backpropagation learning: the RPROP algorithm](https://doi.org/10.1109/ICNN.1993.298623)

There are two variants implemented in this contribution:

1. iRprop+ which is described in the article
  [Empirical evaluation of the improved Rprop learning algorithms](https://doi.org/10.1016/S0925-2312(01)00700-7)

2. Rprop-  which is described in Riedmiller's article
  [Advanced supervised learning in multi-layer perceptrons — From backpropagation to adaptive learning algorithms](https://doi.org/10.1016/0920-5489(94)90017-5), and is referred to as Rprop- (Rprop without weight-backtracking) in the
  article [Empirical evaluation of the improved Rprop learning algorithms](https://doi.org/10.1016/S0925-2312(01)00700-7).




The TensorFlow implementation is described in the article
  [Resilient Backpropagation (Rprop) for Batch-learning in TensorFlow](https://openreview.net/forum?id=r1R0o7yDz)

ciprianflow committed:
Python impl, tests, docs

ciprianflow committed:
Implementation for the CPU and GPU kernels, tests

christian-igel committed:
revision of documentation

christian-igel committed:
revision of documentation

ciprianflow committed:
docs cleanup and typos

ciprianflow committed:
added resource and cleanup

ciprianflow committed:
Merge conflict resolved in contrib/opt/ini.py

christian-igel commented:
I confirm that I am OK with the pull request and with my commits being contributed to this project.
Cheers,
Christian

rmlarsen commented:
@ciprianflow Please sign the CLA.

ciprianflow commented:
@googlebot rescan

ciprianflow commented:
I signed it!

rmlarsen commented:
@ciprianflow is it possible that you signed it with a different email than the one(s) used to commit the code in the PR? The CLA bot seems to indicate that.

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus.py:
```diff
@@ -0,0 +1,319 @@
+  """
+
+  # Values for gate_gradients, taken from base class.
+  GATE_NONE = 0
```
Don't repeat these, just reuse Optimizer.GATE etc

ciprianflow replied:
ok

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus.py:
```diff
@@ -0,0 +1,319 @@
+    self._delta_max_t = None
+
+    # Error tensors
+    self._error = variable_scope.variable(0.0,
```
Defer creation of tensors or variables to prepare, since optimizers are supposed to be reusable across multiple graphs.

Also use _create_non_slot_variable.

I'd prefer it if you inherited from OptimizerV2 instead of Optimizer as we'll do the switch soon.

ciprianflow replied:
I've inherited from OptimizerV2 and create the error variables using  _create_non_slot_variable. However we still need a variable that can hold the loss value at the current iteration before calling apply_gradients, in order to feed the error in the update step. That is also because _create_vars is called in apply_gradients method.

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus.py:
```diff
@@ -0,0 +1,319 @@
+        grad, use_locking=self._use_locking)
+
+
+  def minimize(self, loss, global_step=None, var_list=None,
```
I don't think you need to overload minimize

ciprianflow replied:
The iRprop+ optimizer is dependent on the loss value for the update step. This value is needed for the computation in every iteration. This is the part which makes implementing RProp non-trivial.

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus.py:
```diff
@@ -0,0 +1,319 @@
+    return control_flow_ops.group(*update_ops + [old_error_update],
+                                  name=name_scope)
+
+  def apply_gradients(self, grads_and_vars, loss, global_step=None, name=None):
```
I don't think you need to overload apply_gradients

ciprianflow replied:
Same comment as above.

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus_test.py:
```diff
@@ -0,0 +1,272 @@
+      # type error in case no loss variable is passed, sanity check
+      self.assertRaises(TypeError, opt.apply_gradients, grads)
+
+  def testDense(self):
```
Decorate this test with test_util.run_in_graph_and_eager_modes; as it is now it has a bunch of ifs related to eager which will never execute

ciprianflow replied:
ok

alextp commented on tensorflow/contrib/opt/python/training/rprop_minus_test.py:
```diff
@@ -0,0 +1,180 @@
+          self.assertAllCloseAccordingToType(var0_np, self.evaluate(var0))
+          self.assertAllCloseAccordingToType(var1_np, self.evaluate(var1))
+
+  def testDense(self):
```
Same comments as the other test

alextp commented on tensorflow/core/kernels/training_ops_gpu.cu.cc:
```diff
@@ -284,6 +340,61 @@ struct ApplyPowerSign<GPUDevice, T> {
+                  typename TTypes<T>::ConstFlat grad) {
+    const int32 N = var.size();
+
+    CudaLaunchConfig config = GetCudaLaunchConfig(N, d);
```
Why does this optimizer need this cuda stuff while other optimizers don't?

ciprianflow replied:
It is not possible to implement the update step of RProp using the GPU implementation (of the operations) of TensorFlow and Eigen-Tensor in a trivial or efficient way, therefore we use CudaLaunchConfig to determine the launch parameters (of the GPU hardware) in order to execute the CUDA kernel that computes the update step.

alextp replied:
Can you add this as a comment to the code, so future readers know what is going on? (specifically what eigen doesn't let you do would be very helpful)

alextp commented on tensorflow/core/kernels/training_ops_test.cc:
```diff
@@ -308,4 +308,92 @@ static void BM_PowerSign(int iters, int params) {
 }
 BENCHMARK(BM_PowerSign)->Arg(128 << 10)->Arg(256 << 10);
 
+static void IRpropPlus(int32 n, Graph** init_g, Graph** train_g) {
```
I think the python tests are sufficient for these optimizers

ciprianflow replied:
removed

ciprianflow committed:
Rprop now inherits from OptimizerV2, updated the tests to include eager execution

ciprianflow committed:
removed benchmark for Rprop

ciprianflow commented:
Thank you for the feedback @alextp
I have responded and pushed the changes.

@rmlarsen I think we both signed the agreements. I've checked the email used for the commits and it is the same. Trying again..
@googlebot rescan

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus.py:
```diff
@@ -0,0 +1,284 @@
+
+    error, old_error = self._get_error_values(state)
+    # Update the error E(t) passed in apply_gradients or minimize
+    update_error = error.assign(self._error, use_locking=self._use_locking)
```
This update is getting dropped on the floor because nothing depends on it. Add a control dependency to it to make sure it executes.

ciprianflow replied:
ok

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus.py:
```diff
@@ -0,0 +1,284 @@
+    delta_update = state.get_slot(var, "delta_update")
+
+    error, old_error = self._get_error_values(state)
+    update_error = error.assign(
```
Same issue here

ciprianflow replied:
ok

alextp commented on tensorflow/contrib/opt/python/training/irprop_plus.py:
```diff
@@ -0,0 +1,284 @@
+
+    # Error E(t)
+    if not self._is_scalar(loss):
+      raise ValueError("'loss' (%s) must be a 0-D tensor." % loss)
```
Can you drop these scalar tests? I think they are redundant / not present in other optimizers. TF will silently reduce non-scalar losses to scalars I think. If you drop this you can stop overriding apply_gradients and minimize and your code will be shorter and easier to read and more future-proof.

ciprianflow replied:
I've tested this part and TF does not reduce the loss to scalar. Although, not providing a reduce op for the loss does work for other optimizers, for iRprop+ it is required.
An alternative would be to provide a default reduce operation in case none is provided. What do you think?

alextp replied:
tf.gradients implicitly reduces the loss to a scalar so you do not need this.

ciprianflow replied:
@alextp It does not reduce the loss to a scalar for me (tested in OptimizerV2), unless the weights are scalars. For example, after running the code below, the cost is a vector which needs to be reduced to a scalar in order to feed the error to IRprop+.  Am I missing something? some help would be much appreciated. Thank you.

import tensorflow as tf

x1 = tf.Variable([1.0, 2.0])
x2 = tf.Variable([2.0, 3.0])
loss = 2*x1**2 + 3*x2

grad = tf.gradients(loss, [x1, x2])

init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    grad_value = sess.run(grad)
    print(sess.run(loss))

alextp commented on tensorflow/core/kernels/training_ops.cc:
```diff
@@ -418,6 +418,81 @@ struct ApplyPowerSign<CPUDevice, T> {
+        // revert deltaW(t-1) first then update delta
+        if (error() > old_error())
+          // compute deltaW and update the weight
+          var(i) -= delta_update(i) * (-sgn(old_grad(i)));
```
I think this should be var.device(i) -= ... (same below)

ciprianflow replied:
the code is correct (var is a Flat tensor and each coefficient is updated individually according to the sign of the gradients), the implementation for cpu is not multi threaded.

alextp commented on tensorflow/core/kernels/training_ops.cc:
```diff
@@ -3664,4 +3739,282 @@ REGISTER_KERNELS(GPU, double);
+
+}  // namespace functor
+
+REGISTER_KERNELS(GPU, Eigen::half);
```
Replace with TF_CALL_GPU_NUMBER_TYPES?

ciprianflow replied:
ok

ciprianflow committed:
Added control dependencies to non slot variables and some comments for getCudaLaunchConfig

asimshankar commented on tensorflow/core/ops/training_ops.cc:
```diff
@@ -1053,4 +1053,94 @@ REGISTER_OP("ResourceApplyPowerSign")
       return ApplyPowerSignShapeFn(c, /*sparse=*/false);
     });
 
+static Status ApplyRpropShapeFn(InferenceContext* c, bool withError) {
```
(For tf-api-owners)

The op and kernel should also live in tensorflow/contrib instead of in tensorflow/core?

tensorflowbutler commented:
Nagging Assignee @rmlarsen: It has been 28 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @rmlarsen: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

alextp commented:
Sorry if I wasn't clear. Passing a non-scalar loss to tf.gradients will
implicitly reduce it to a scalar (though the scalar is never actually
computed). Since your code doesn't seem to use the non-scalar loss in any
way other than to pass it to tf.gradients I believe you do not need to
reduce it.

On Mon, Sep 17, 2018 at 5:15 PM Ciprian F. <notifications@github.com> wrote:

> *@ciprianflow* commented on this pull request.
> ------------------------------
>
> In tensorflow/contrib/opt/python/training/irprop_plus.py
> <https://github.com/tensorflow/tensorflow/pull/20918#discussion_r218266054>
> :
>
> > +    @compatibility(eager)
> +    When eager execution is enabled, `loss` should be a Python function that
> +    takes elements of `var_list` as arguments and computes the value to be
> +    minimized. If `var_list` is `None`, `loss` should take no arguments.
> +    Minimization (and gradient computation) is done with respect to the
> +    elements of `var_list` if not `None`, else with respect to any trainable
> +    variables created during the execution of the `loss` function.
> +    `gate_gradients`, `aggregation_method`, `colocate_gradients_with_ops` and
> +    `grad_loss` are ignored when eager execution is enabled.
> +    @end_compatibility
> +    """
> +    # Override method from base class, the loss is required to be scalar
> +
> +    # Error E(t)
> +    if not self._is_scalar(loss):
> +      raise ValueError("'loss' (%s) must be a 0-D tensor." % loss)
>
> @alextp <https://github.com/alextp> It does not reduce the loss to a
> scalar for me (tested in OptimizerV2), unless the weights are scalars. For
> example, after running the code below, the cost is a vector which needs to
> be reduced to a scalar in order to feed the error to IRprop+. Am I missing
> something? some help would be much appreciated. Thank you.
>
> import tensorflow as tf
>
> x1 = tf.Variable([1.0, 2.0])
> x2 = tf.Variable([2.0, 3.0])
> loss = 2*x1**2 + 3*x2
>
> grad = tf.gradients(loss, [x1, x2])
>
> init = tf.global_variables_initializer()
>
> with tf.Session() as sess:
> sess.run(init)
> grad_value = sess.run(grad)
> print(sess.run(loss))
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/20918#discussion_r218266054>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxcPk9s1sDpUbJw3vebngZzbv14qqks5ucDsogaJpZM4VUV9I>
> .
>


-- 
 - Alex

rmlarsen commented:
@ciprianflow Could you please rebase you PR to resolve merge conflicts?

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

christian-igel commented:
> It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

Yes, the PR is still valid. Sorry for the long silence, we will work on it asap.

tensorflowbutler commented:
It has been 28 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
